File: modules/offline-core-android/src/main/kotlin/cc/musicme/offline/LicenseClaims.kt
package cc.musicme.offline

import okio.ByteString.Companion.decodeBase64
import org.json.JSONObject
import java.nio.charset.StandardCharsets

data class LicenseClaims(
    val trackId: String,
    val mid: Long,
    val deviceId: String,
    val userId: String,
    val key: ByteArray,     // 32 bytes AES-256
    val iv: ByteArray,      // 16 bytes AES-CTR base IV
    val exp: Long,          // unix seconds
    val iat: Long,
    val v: String
) {
    companion object {
        const val SUPPORTED_VERSION = "offline-v1"

        fun decode(jwt: String): LicenseClaims {
            val parts = jwt.split(".")
            if (parts.size != 3) throw OfflineError.MalformedLicense

            val bodyBytes = try {
                base64UrlDecode(parts[1])
            } catch (e: IllegalArgumentException) {
                throw OfflineError.MalformedLicense
            }
            val json = try {
                JSONObject(String(bodyBytes, StandardCharsets.UTF_8))
            } catch (e: Exception) {
                throw OfflineError.MalformedLicense
            }

            try {
                val trackId = json.getString("trackId")
                val mid = json.getLong("mid")
                val deviceId = json.getString("deviceId")
                val userId = json.getString("userId")
                val keyB64 = json.getString("key")
                val ivB64 = json.getString("iv")
                val exp = json.getLong("exp")
                val iat = json.getLong("iat")
                val v = json.getString("v")

                val key = keyB64.decodeBase64()?.toByteArray()
                    ?: throw OfflineError.MalformedLicense
                val iv = ivB64.decodeBase64()?.toByteArray()
                    ?: throw OfflineError.MalformedLicense
                if (key.size != 32) throw OfflineError.MalformedLicense
                if (iv.size != 16) throw OfflineError.MalformedLicense
                if (v != SUPPORTED_VERSION) throw OfflineError.UnsupportedLicenseVersion

                return LicenseClaims(trackId, mid, deviceId, userId, key, iv, exp, iat, v)
            } catch (e: OfflineError) {
                throw e
            } catch (e: Exception) {
                throw OfflineError.MalformedLicense
            }
        }
    }
}

// okio's decodeBase64 accepts both the standard and url-safe alphabets, padded
// or not — and unlike java.util.Base64 it exists on every Android API level
// (java.util.Base64 is API 26+; the module supports minSdk 24 / Android 7).
internal fun base64UrlDecode(input: String): ByteArray =
    input.decodeBase64()?.toByteArray() ?: throw IllegalArgumentException("invalid base64: $input")

File: demos/react-native/modules/offline/ios/OfflineCore/LicenseClaims.swift
import Foundation

public struct LicenseClaims: Equatable {
    public let trackId: String
    public let mid: Int64
    public let deviceId: String
    public let userId: String
    public let key: Data    // 32 bytes (AES-256)
    public let iv: Data     // 16 bytes (AES-CTR base IV)
    public let exp: Int64   // unix seconds
    public let iat: Int64
    public let v: String

    public static let supportedVersion = "offline-v1"

    public static func decode(jwt: String) throws -> LicenseClaims {
        let parts = jwt.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3 else { throw OfflineError.malformedLicense }

        let body = String(parts[1])
        guard let bodyData = base64URLDecode(body),
              let json = try? JSONSerialization.jsonObject(with: bodyData) as? [String: Any]
        else { throw OfflineError.malformedLicense }

        guard
            let trackId = json["trackId"] as? String,
            let mid = (json["mid"] as? Int).map(Int64.init) ?? (json["mid"] as? Int64),
            let deviceId = json["deviceId"] as? String,
            let userId = json["userId"] as? String,
            let keyB64 = json["key"] as? String,
            let ivB64 = json["iv"] as? String,
            let exp = (json["exp"] as? Int).map(Int64.init) ?? (json["exp"] as? Int64),
            let iat = (json["iat"] as? Int).map(Int64.init) ?? (json["iat"] as? Int64),
            let v = json["v"] as? String,
            let key = Data(base64Encoded: keyB64),
            let iv = Data(base64Encoded: ivB64)
        else { throw OfflineError.malformedLicense }

        guard key.count == 32 else { throw OfflineError.malformedLicense }
        guard iv.count == 16 else { throw OfflineError.malformedLicense }
        guard v == supportedVersion else { throw OfflineError.unsupportedLicenseVersion }

        return LicenseClaims(
            trackId: trackId, mid: mid, deviceId: deviceId, userId: userId,
            key: key, iv: iv, exp: exp, iat: iat, v: v
        )
    }
}

func base64URLDecode(_ input: String) -> Data? {
    var s = input
        .replacingOccurrences(of: "-", with: "+")
        .replacingOccurrences(of: "_", with: "/")
    let pad = (4 - s.count % 4) % 4
    s.append(String(repeating: "=", count: pad))
    return Data(base64Encoded: s)
}
